Dashicons: Implement with Interactivity API

sirreal opened the pull request:
Replace the [dashicon page](https://developer.wordpress.org/resource/dashicons/) shortcode with a dashicon page block.

I had to replace the shortcode with a block because `wpautop` was running on the shortcode and mangling Interactivity API HTML.

I have commented the problematic line and added a todo comment to re-enabled this when supported. ~This requires https://github.com/WordPress/wordpress-develop/pull/7075 to work correctly on the server. I expect that to be released in 6.6.2. To test this, you'll need to check out that PR and instruct wp-env to use that core version.~


I have not removed or touched the existing dashicon shortcode code for now, but there are opportunities for removing code if we go ahead with this.

Most of the page HTML is transformed from the shortcode in https://github.com/WordPress/wporg-developer/blob/trunk/source/wp-content/themes/wporg-developer-2023/inc/shortcode-dashicons.php.

## Testing

Load up the page in a local environment and confirm it behaves like prodoction. Note that these pages render a random icon in detail:

- https://developer.wordpress.org/resource/dashicons/
- http://localhost:8888/resource/dashicons/

"Permalinks" were rendered client site with a hash to the icon slug. This behavior has been changed to use query params so that the server can render the requested icon (hash is not available to the server). Client-side, if a fragment is found with no valid icon query param, it will be upgraded to a query param:

- http://localhost:8888/resource/dashicons/?icon=dashicons-id-alt

See the upgrade behavior, where these will behave the same (after client rewrites the URL):
- https://developer.wordpress.org/resource/dashicons/#id-alt
- http://localhost:8888/resource/dashicons/#id-alt

Filtering should work basically the same. Try using the filter bar. Non-alphanumeric characters are ignored from filtering.

The copy buttons work the same, but they are buttons instead of `a` elements. This seems more semantic as they open a prompt.

sirreal committed:
Add experimental-modules build flag

sirreal committed:
Add experimental-modules build flag

sirreal committed:
Try making a block

sirreal committed:
Try making a block

sirreal committed:
PICKME: Upgrade @wordpress/scripts

sirreal committed:
Not rendering a post now

sirreal committed:
Not rendering a post now

sirreal committed:
starting to work properly

sirreal committed:
starting to work properly

sirreal committed:
Close to working

sirreal committed:
Close to working

sirreal committed:
Remove redundant property

sirreal committed:
Remove redundant property

sirreal committed:
Copy buttons working

sirreal committed:
Copy buttons working

sirreal committed:
Remove unused context

sirreal committed:
Remove unused context

sirreal committed:
Localize copy texts

sirreal committed:
Localize copy texts

sirreal committed:
Add icon to url query param when selecting

sirreal committed:
Add icon to url query param when selecting

sirreal committed:
Revert "Add icon to url query param when selecting"

This reverts commit a04fd43f4a600744a29076dfcb24e541cfdb415e.

sirreal committed:
Revert "Add icon to url query param when selecting"

This reverts commit a04fd43f4a600744a29076dfcb24e541cfdb415e.

sirreal committed:
Filter icons

sirreal committed:
Filter icons

sirreal committed:
Handle updating URL for selected icon and moving to page top

sirreal committed:
Handle updating URL for selected icon and moving to page top

sirreal committed:
Handle icon query param to select icon

sirreal committed:
Handle icon query param to select icon

sirreal committed:
Upgrade fragment-style links to query param

sirreal committed:
Upgrade fragment-style links to query param

sirreal commented on source/wp-content/themes/wporg-developer-2023/package.json:
```diff
@@ -16,8 +16,8 @@
-		"build": "wp-scripts build",
-		"start": "wp-scripts start",
+		"build": "wp-scripts build --experimental-modules",
+		"start": "wp-scripts start --experimental-modules",
```
I could pull this out to its own PR to upgrade the package and add this flag. I don't think it should affect anything else.

sirreal committed:
Disable server-side intermediate derived state

See https://github.com/WordPress/wordpress-develop/pull/7075

sirreal committed:
Disable server-side intermediate derived state

See https://github.com/WordPress/wordpress-develop/pull/7075

sirreal committed:
Remove unused block properties

sirreal committed:
Remove unused block properties

sirreal committed:
Remove console.log

sirreal committed:
Remove console.log

sirreal commented:
@ryelle I'd appreciate feedback on this PR, thank you!

gziolo commented on package.json:
```diff
@@ -6,7 +6,7 @@
 	"private": true,
 	"dependencies": {
-		"@wordpress/scripts": "27.1.0"
+		"@wordpress/scripts": "28.3.0"
```
I assume that the code is going to work with WordPress 6.7 because otherwise, it would fail with the missing JSX runtime script dependency.

sirreal replied:
Good callout!

[The v28 changelog reads:](https://github.com/WordPress/gutenberg/blob/eea7183c6b156610cd3daa538e98a8b4b5388c2e/packages/scripts/CHANGELOG.md#2800-2024-05-31)

> **Breaking Changes**
> Note If you're using @wordpress/scripts for building JS scripts to target WordPress 6.5 or earlier, you should not upgrade to this version and continue using @wordpress/scripts@27.

I think the site uses latest WordPress, so if it's on WP 6.6, it should work if that comment is accurate. Do you know if it's WordPress 6.6 or 6.7 that will use the new JSX runtime?

@ryelle can you confirm the site runs on latest WordPress release (6.6.x at this time)?

gziolo replied:
Right, 6.6 is fine, too. I assume that it's sartisfied here. However, I wanted to ensure we don't see any surprises when deploying.

ryelle replied:
WordPress.org runs trunk & (usually) the latest Gutenberg version (the exception is the Support Forums due to conflicts with [blocks-everywhere](https://github.com/Automattic/blocks-everywhere), so we used a pinned version there). So we should be set with this version change. It also worked fine on my sandbox, which matches production versions.

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/dashicons-page/render.php:
```diff
@@ -0,0 +1,262 @@
+}
+
+wp_interactivity_config(
+	'wporg/dashicons-page',
```
When defining interactivity stores in themes, I've been naming them with the theme namespace too, just to differentiate from the shared interactive blocks in [wporg-mu-plugins](https://github.com/WordPress/wporg-mu-plugins/tree/trunk/mu-plugins/blocks). Would you mind updating the name to `wporg/developer/dashicons-page`?

sirreal replied:
I changed this for the store. I did not modify the block name, let me know if you'd like me to use another name for the block.

ryelle replied:
Nope, the block name is fine, just the stores. Thanks!

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/dashicons-page/render.php:
```diff
@@ -0,0 +1,262 @@
+	get_stylesheet_directory_uri() . '/stylesheets/page-dashicons.css',
+	array(),
+	get_stylesheet_directory() . '/stylesheets/page-dashicons.css'
+);
```
You could also move this stylesheet into the block, to keep it all in one place— once this page is updated to use the block, the old shortcode & related code can be removed.

sirreal replied:
I moved this into the block's `viewStyle`. As a follow-up when removing dashicons shortcode, the stylesheet can move into the block source.

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/dashicons-page/render.php:
```diff
@@ -0,0 +1,262 @@
+	'dashicons-page',
+	get_stylesheet_directory_uri() . '/stylesheets/page-dashicons.css',
+	array(),
+	get_stylesheet_directory() . '/stylesheets/page-dashicons.css'
```
This should be a version number, the previous code used the modified time:

```suggestion
	filemtime( get_stylesheet_directory() . '/stylesheets/page-dashicons.css' )
```

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/dashicons-page/render.php:
```diff
@@ -0,0 +1,262 @@
+			'copyCss' => _( 'Copy this, then paste in your CSS :before selector.' ),
+			'copyHtml' =>  _( 'Copy this, then paste in your HTML.' ),
+			'copyGlyph' => _( 'Copy this, then paste in your Photoshop textfield.' ),
+		),
```
These strings should be double-underscore, and have a text domain: `__( 'Copy …', 'wporg' )`

ryelle reviewed (commented):
Overall this is working just as the current page 👍🏻  I appreciate that you've kept the old URLs working, and the server-side rendering of the icon param works nicely.

There are a handful of lint errors, I know you mentioned not having it set up locally — I don't know if this helps, but this is the output at the moment (ignoring some content escaping issues that exist on the shortcode too).

```
$ composer lint ./source/wp-content/themes/wporg-developer-2023/src/dashicons-page
> phpcs './source/wp-content/themes/wporg-developer-2023/src/dashicons-page'
.E 2 / 2 (100%)

FILE: ./source/wp-content/themes/wporg-developer-2023/src/dashicons-page/render.php
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 11 ERRORS AFFECTING 9 LINES
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  26 | ERROR | [x] Array keys must be surrounded by spaces unless they contain a string or an integer. (WordPress.Arrays.ArrayKeySpacingRestrictions.NoSpacesAroundArrayKeys)
  27 | ERROR | [x] Array keys must be surrounded by spaces unless they contain a string or an integer. (WordPress.Arrays.ArrayKeySpacingRestrictions.NoSpacesAroundArrayKeys)
  33 | ERROR | [ ] $_GET data not unslashed before sanitization. Use wp_unslash() or similar (WordPress.Security.ValidatedSanitizedInput.MissingUnslash)
  33 | ERROR | [ ] Detected usage of a non-sanitized input variable: $_GET['icon'] (WordPress.Security.ValidatedSanitizedInput.InputNotSanitized)
  42 | ERROR | [ ] Found single-underscore "_()" function when double-underscore expected. (WordPress.WP.I18n.SingleUnderscoreGetTextFunction)
  43 | ERROR | [x] Expected 1 space after "=>"; 2 found (WordPress.WhiteSpace.OperatorSpacing.SpacingAfter)
  43 | ERROR | [ ] Found single-underscore "_()" function when double-underscore expected. (WordPress.WP.I18n.SingleUnderscoreGetTextFunction)
  44 | ERROR | [ ] Found single-underscore "_()" function when double-underscore expected. (WordPress.WP.I18n.SingleUnderscoreGetTextFunction)
  53 | ERROR | [ ] Empty line required before block comment (Squiz.Commenting.BlockComment.NoEmptyLineBefore)
  77 | ERROR | [x] Expected 1 space before comment text but found 2; use block comment if you need indentation (Squiz.Commenting.InlineComment.SpacingBefore)
 158 | ERROR | [x] Opening PHP tag must be on a line by itself (Squiz.PHP.EmbeddedPhp.ContentBeforeOpen)
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 5 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

sirreal committed:
Fix incorrect asset version argument, use filemtime

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

sirreal committed:
Fix incorrect asset version argument, use filemtime

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

sirreal committed:
Fix dirname levels lint

sirreal committed:
Fix dirname levels lint

sirreal committed:
Fix translation function __ and textdomain wporg

sirreal committed:
Fix translation function __ and textdomain wporg

sirreal committed:
Update store namespace to wporg/developer/dashicons-page

sirreal committed:
Update store namespace to wporg/developer/dashicons-page

sirreal committed:
Fix =,=> alignment lints

sirreal committed:
Fix =,=> alignment lints

sirreal committed:
Fix whitespace lints

sirreal committed:
Fix whitespace lints

sirreal committed:
Move dashicons page css into block viewStyle

sirreal committed:
Move dashicons page css into block viewStyle

sirreal commented:
I've addressed feedback and fixed lints, this is ready for another review.

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/dashicons-page/view.js:
```diff
@@ -0,0 +1,104 @@
+		},
+
+		get style() {
+			const cleanedValue = state.filter.replace( /[^a-zA-Z0-9]/, '' );
```
This only replaces the first matching instance, so `s.ite` works but `s.i.t.e` doesn't.

I would also suggest allowing `-`, since some icons use that as part of the name. For example, [menu-alt](https://developer.wordpress.org/resource/dashicons/?icon=dashicons-menu-alt) can't be found by searching "menu-alt".

```suggestion
			const cleanedValue = state.filter.replace( /[^a-zA-Z0-9-]/g, '' );
```

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/dashicons-page/render.php:
```diff
@@ -0,0 +1,255 @@
+	$icons_sections[] = $icon_section;
+}
+
+$selected_icon = $_GET['icon'] ?? '';
```
This line flags some security lint issues, but it's OK since the value is matched against the icon list. I'd add a comment disabling the check to avoid those false errors.

```suggestion
$selected_icon = $_GET['icon'] ?? ''; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput -- Value checked against icon safelist.
```

sirreal committed:
Fix replace regex to run global and allow -

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

sirreal committed:
Fix replace regex to run global and allow -

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

sirreal committed:
Add ignore comment on valid user input usage

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

sirreal committed:
Add ignore comment on valid user input usage

Co-authored-by: Kelly Dwan <ryelle@users.noreply.github.com>

sirreal committed:
Restore server-side eachIcon derived state

Core supports derived state in non-final position since r58825, so the
icon list and detail panel are now fully rendered on the server.

sirreal committed:
Guard against non-string icon query argument

$_GET['icon'] can be an array (?icon[]=x), which made array_key_exists()
throw a TypeError and render the critical error page.

sirreal committed:
Make icon list items keyboard accessible

Use a button for each icon so it is focusable and activatable from the
keyboard.

sirreal committed:
Scroll to the glyph on every icon selection

Setting location.hash to its current value is a no-op, so only the first
click scrolled. Use scrollIntoView instead, which also avoids pushing a
history entry.

sirreal committed:
Update wp-class-example with the selected icon

Restores behavior from the previous implementation.

sirreal commented:
Abandoning although there was good work and potential here.
